README.md
# cypress-plugin-visual-tests
> Plugin for snapshot tests in [Cypress.io](https://www.cypress.io/).

> cypress-plugin-visual-tests is a fork of cypress-plugin-snapshots.

[![NPM][npm-icon]][npm-url]

[![Build status][ci-image]][ci-url]

![cypress-plugin-visual-tests](docs/images/cypress-plugin-visual-tests.png)

## Installation
`npm i cypress-plugin-visual-tests -S`

## Usage for text snapshots
```javascript
describe('data test', () => {
  it('toMatchSnapshot - JSON', () => {
    return cy.request('data.json')
      .its('body')
      .toMatchSnapshot();
  });

  it('toMatchSnapshot - JSON with options', () => {
    return cy.request('data.json')
      .its('body')
      .toMatchSnapshot({
        ignoreExtraFields: true,
      });
  });

  it('toMatchSnapshot - HTML', () => {
    cy.visit('page.html')
      .then(() => {
        cy.get('div').toMatchSnapshot();
      });
  });
});
```

You can pass the following options to `toMatchSnapshot` to override default behavior.
```javascript
{
  "ignoreExtraFields": false,         // Ignore fields that are not in snapshot
  "ignoreExtraArrayItems": false,     // Ignore if there are extra array items in result
  "normalizeJson": true,              // Alphabetically sort keys in JSON
  "replace": {                        // Replace `${key}` in snapshot with `value`.
    "key": "value",
  }
}
```

**`replace`**
Use `replace` with caution. Tests should be deterministic. It's often a better solution to influence your
test result instead of your snapshot (by mocking data for example).

## Usage for image snapshots
```javascript
it('toMatchImageSnapshot - element', () => {
  cy.visit('/static/stub.html')
    .then(() => {
      cy.get('[data-test=test]')
        .toMatchImageSnapshot();
    });
});

it('toMatchImageSnapshot - whole page', () => {
  cy.visit('/static/stub.html')
    .then(() => {
      cy.document()
        .toMatchImageSnapshot();
    });
});
```

You can pass the following options to `toMatchImageSnapshot` to override default behavior.
```javascript
{
  "imageConfig": {
    "createDiffImage": true,                // Should a "diff image" be created, can be disabled for performance
    "threshold": 0.01,                      // Amount in pixels or percentage before snapshot image is invalid
    "thresholdType": "percent",             // Can be either "pixel" or "percent"
  },
  "name": "custom image name",            // Naming resulting image file with a custom name rather than concatenating test titles
  "separator": "custom image separator",  // Naming resulting image file with a custom separator rather than using the default ` #`
}
```

You can also use any option from the `cypress.screenshot` [arguments list](https://docs.cypress.io/api/commands/screenshot.html#Arguments).

For example:
```javascript
cy.get('.element')
  .toMatchImageSnapshot({
    clip: { x: 0, y: 0, width: 100, height: 100 },
  });
```

## Configure Cypress.io
Add this to your `cypress.json` configuration file:
```json
"ignoreTestFiles": [
  "**/__snapshots__/*",
  "**/__image_snapshots__/*"
]
```

### Plugin
Find your `cypress/plugins/index.js` file and change it to look like this:

```javascript
const { initPlugin } = require('cypress-plugin-visual-tests/plugin');

module.exports = (on, config) => {
  initPlugin(on, config);
  return config;
};
```

### Command
Find your `cypress/support/index.js` file and add the following line:

```javascript
import 'cypress-plugin-visual-tests/commands';
```

### Make changes to default configuration
You can customize the configuration in the `cypress.json` file in the root of your Cypress project.

Add the configuration below to your `cypress.json` file to make changes to the default values.

```javascript
"env": {
  "cypress-plugin-snapshots": {
    "autoCleanUp": false,            // Automatically remove snapshots that are not used by test
    "autopassNewSnapshots": true,    // Automatically save & pass new/non-existing snapshots
    "diffLines": 3,                  // How many lines to include in the diff modal
    "excludeFields": [],             // Array of fieldnames that should be excluded from snapshot
    "ignoreExtraArrayItems": false,  // Ignore if there are extra array items in result
    "ignoreExtraFields": false,      // Ignore extra fields that are not in `snapshot`
    "normalizeJson": true,           // Alphabetically sort keys in JSON
    "prettier": true,                // Enable `prettier` for formatting HTML before comparison
    "imageConfig": {
      "createDiffImage": true,       // Should a "diff image" be created, can be disabled for performance
      "resizeDevicePixelRatio": true,// Resize image to base resolution when Cypress is running on high DPI screen, `cypress run` always runs on base resolution
      "threshold": 0.01,             // Amount in pixels or percentage before snapshot image is invalid
      "thresholdType": "percent"     // Can be either "pixels" or "percent"
    },
    "screenshotConfig": {            // See https://docs.cypress.io/api/commands/screenshot.html#Arguments
      "blackout": [],
      "capture": "fullPage",
      "clip": null,
      "disableTimersAndAnimations": true,
      "log": false,
      "scale": false,
      "timeout": 30000
    },
    "serverEnabled": true,           // Enable "update snapshot" server and button in diff modal
    "serverHost": "localhost",       // Hostname for "update snapshot server"
    "serverPort": 2121,              // Port number for  "update snapshot server"
    "updateSnapshots": false,        // Automatically update snapshots, useful if you have lots of changes
    "backgroundBlend": "difference"  // background-blend-mode for diff image, useful to switch to "overlay"
  }
}
```

## Caveats :warning:
There is currently an issue when running "All Tests" in Cypress with this plugin. You can follow the progress on the issue [here](https://github.com/meinaart/cypress-plugin-snapshots/issues/10) and [here](https://github.com/cypress-io/cypress/issues/3090). When running "All Tests" any tests that utilize `cypress-plugin-visual-tests` will throw an error.

## Roadmap
Below is a list of functionality that is under consideration for implementing in a next version.

- Fix handling of "update snapshot" button that contains a replacable field
- Disable "update snapshots" server in headless mode
- Add more unit tests
- Add [JSDoc](http://usejsdoc.org/) documentation
- Improve TypeScript bindings

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style.

## License
This plugin is released under the MIT license.

[npm-icon]: https://nodei.co/npm/@kiru42/cypress-plugin-visual-tests.svg?downloads=true
[npm-url]: https://www.npmjs.com/package/@kiru42/cypress-plugin-visual-tests
[ci-image]: https://github.com/kiru42/cypress-plugin-visual-tests/workflows/CI/badge.svg?branch=main
[ci-url]: https://github.com/kiru42/cypress-plugin-visual-tests/actions


CHANGELOG.md
## [1.5.37](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.36...v1.5.37) (2023-12-22)


### Bug Fixes

* **deps:** update dependency jimp to v0.22.10 ([13d7da3](https://github.com/kiru42/cypress-plugin-visual-tests/commit/13d7da3cf7ef46b1d0a91c73d16af1ca36b365ca))

## [1.5.36](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.35...v1.5.36) (2023-05-07)


### Bug Fixes

* **deps:** update dependency diff2html to v3.4.35 ([884d84d](https://github.com/kiru42/cypress-plugin-visual-tests/commit/884d84dcbf03c330c455f1926b41e9e7a9fb15f3))
* **deps:** update socket.io packages to v4.6.1 ([2ac48d4](https://github.com/kiru42/cypress-plugin-visual-tests/commit/2ac48d46068f4759a926e454768870b9587141dc))

## [1.5.35](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.34...v1.5.35) (2023-05-07)


### Bug Fixes

* **deps:** update dependency minimist to v1.2.8 ([5389961](https://github.com/kiru42/cypress-plugin-visual-tests/commit/53899612937c0297db5019d7dac4403c5d58e8cd))

## [1.5.34](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.33...v1.5.34) (2023-05-06)


### Bug Fixes

* **deps:** update dependency express to v4.18.2 ([f3368e0](https://github.com/kiru42/cypress-plugin-visual-tests/commit/f3368e0ed02bf66b24656b6137a2a952afe086fe))

## [1.5.33](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.32...v1.5.33) (2022-09-28)


### Bug Fixes

* **deps:** update dependency jimp to v0.16.2 ([058c7f7](https://github.com/kiru42/cypress-plugin-visual-tests/commit/058c7f7eab29771042ebe1eafa293aa0bcf94bde))
* **deps:** update dependency ws to v8.9.0 ([10a4668](https://github.com/kiru42/cypress-plugin-visual-tests/commit/10a46681af7e9a5362644b9dab08d3f21b2c7d94))

## [1.5.32](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.31...v1.5.32) (2022-09-03)


### Bug Fixes

* **deps:** update dependency diff2html to v3.4.19 ([e6a3768](https://github.com/kiru42/cypress-plugin-visual-tests/commit/e6a3768149462092e562c066350be7b634d869f5))
* **deps:** update socket.io packages to v4.5.2 ([311e7fe](https://github.com/kiru42/cypress-plugin-visual-tests/commit/311e7fee85a35f517ddc0ed5f7e78609db4b6e26))

## [1.5.31](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.30...v1.5.31) (2022-08-12)


### Bug Fixes

* **deps:** update dependency diff2html to v3.4.18 ([7fb486b](https://github.com/kiru42/cypress-plugin-visual-tests/commit/7fb486b376c5cb395b0c7b20a90d843694e5cb93))
* **deps:** update dependency image-size to v1.0.2 ([bd7de26](https://github.com/kiru42/cypress-plugin-visual-tests/commit/bd7de262647464db760349367faeaf86a446cb54))

## [1.5.30](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.29...v1.5.30) (2022-08-12)


### Bug Fixes

* **deps:** update dependency express to v4.18.1 ([b654901](https://github.com/kiru42/cypress-plugin-visual-tests/commit/b654901eacee1072196deb517259535ae67e58d7))
* **deps:** update dependency ws to v8.8.1 ([27bdc51](https://github.com/kiru42/cypress-plugin-visual-tests/commit/27bdc51dad757656ea5ec31287e100b09ae59406))

## [1.5.29](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.28...v1.5.29) (2022-06-01)


### Bug Fixes

* **deps:** update dependency cypress to v9.7.0 ([f55a402](https://github.com/kiru42/cypress-plugin-visual-tests/commit/f55a4021247526d79b0ecb77deca6abd6c943ecc))
* **deps:** update dependency ws to v8.7.0 ([639e9a3](https://github.com/kiru42/cypress-plugin-visual-tests/commit/639e9a3585d6dd011807dcef629b11c93dbc7a4b))
* **deps:** update socket.io packages to v4.5.1 ([d80879f](https://github.com/kiru42/cypress-plugin-visual-tests/commit/d80879f7c7dafdb9f05135140338eaff98a13a3c))

## [1.5.28](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.27...v1.5.28) (2022-05-01)


### Bug Fixes

* **deps:** update dependency cypress to v9.6.0 ([0c250a7](https://github.com/kiru42/cypress-plugin-visual-tests/commit/0c250a7a09c27e2f85c80d2c64a5dbf5821a6d37))
* **deps:** update dependency pixelmatch to v5.3.0 ([01212b3](https://github.com/kiru42/cypress-plugin-visual-tests/commit/01212b33905e1ac45d1bf3a8c404af3c4c90203f))
* **deps:** update dependency ws to v8.6.0 ([74ef86b](https://github.com/kiru42/cypress-plugin-visual-tests/commit/74ef86b11f025b0670db9934c98cc5bbb6ed6614))

## [1.5.27](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.26...v1.5.27) (2022-04-25)


### Bug Fixes

* **deps:** update socket.io packages to v4.5.0 ([1780112](https://github.com/kiru42/cypress-plugin-visual-tests/commit/17801123aa814bf92cec33318d51d4b4f1faa0ac))

## [1.5.26](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.25...v1.5.26) (2022-04-22)


### Bug Fixes

* **deps:** update dependency diff2html to v3.4.17 ([b44272c](https://github.com/kiru42/cypress-plugin-visual-tests/commit/b44272cd0b8b46c18b864b63e1ef6c8f703ef729))

## [1.5.25](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.24...v1.5.25) (2022-04-18)


### Bug Fixes

* **deps:** update dependency cypress to v9.5.4 ([c717aab](https://github.com/kiru42/cypress-plugin-visual-tests/commit/c717aab226e219ecbdf05c501c28bade123cb52d))
* **deps:** update dependency fs-extra to v10.1.0 ([eee3e3b](https://github.com/kiru42/cypress-plugin-visual-tests/commit/eee3e3b6c43327af47d62d7169157805975020b9))

## [1.5.24](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.23...v1.5.24) (2022-04-09)


### Bug Fixes

* **deps:** update dependency minimist to v1.2.6 ([8de04f6](https://github.com/kiru42/cypress-plugin-visual-tests/commit/8de04f6c336722d252345f478df18d001dc2ef38))

## [1.5.23](https://github.com/kiru42/cypress-plugin-visual-tests/compare/v1.5.22...v1.5.23) (2022-04-07)


### Bug Fixes

* **deps:** update dependency cypress to v9.5.3 ([6ecb188](https://github.com/kiru42/cypress-plugin-visual-tests/commit/6ecb1888c9fcdb4acd33c7934e9b772cc271bd5c))


src/config.js
const randtoken = require('rand-token');
const { merge, cloneDeep, clone } = require('lodash');
const { TYPE_JSON } = require('./dataTypes');

function createToken() {
  return randtoken.generate(128);
}

const DEFAULT_SCREENSHOT_CONFIG = Object.freeze({
  blackout: [],
  capture: 'fullPage',
  clip: null,
  padding: null,
  disableTimersAndAnimations: true,
  log: true,
  scale: false,
  timeout: 30000,
});

const DEFAULT_IMAGE_CONFIG = Object.freeze({
  createDiffImage: true,
  resizeDevicePixelRatio: true,
  threshold: 0.1,
  thresholdType: 'percent', // can be 'percent' or 'pixel'
});

const DEFAULT_CONFIG = Object.freeze({
  autoCleanUp: false,
  autopassNewSnapshots: true,
  diffLines: 3,
  excludeFields: [],
  formatJson: true,
  ignoreExtraArrayItems: false,
  ignoreExtraFields: false,
  imageConfig: clone(DEFAULT_IMAGE_CONFIG),
  normalizeJson: true,
  prettier: true,
  prettierConfig: {
    html: {
      parser: 'html',
      tabWidth: 2,
      endOfLine: 'lf',
    },
  },
  screenshotConfig: clone(DEFAULT_SCREENSHOT_CONFIG),
  serverEnabled: true,
  serverHost: 'localhost',
  serverPort: 2121,
  token: createToken(),
  updateSnapshots: false,
  backgroundBlend: 'difference',
  name: '',
});

const CONFIG_KEY = 'cypress-plugin-snapshots';

let config = cloneDeep(DEFAULT_CONFIG);

function initConfig(initialConfig) {
  if (initialConfig) {
    config = merge(config, initialConfig);
  }
  return config;
}

function getConfig() {
  return config;
}

function getImageConfig(options = {}) {
  return Object.keys(DEFAULT_IMAGE_CONFIG)
    .filter(key => options.imageConfig && options.imageConfig[key] !== undefined)
    .reduce((imageConfig, key) => {
      imageConfig[key] = options.imageConfig[key];
      return imageConfig;
    }, merge({}, DEFAULT_IMAGE_CONFIG, getConfig().imageConfig));
}

function getScreenshotConfig(options = {}) {
  const screenshotConfig = Object.keys(DEFAULT_SCREENSHOT_CONFIG)
    .filter(key => options.screenshotConfig && options.screenshotConfig[key] !== undefined)
    .reduce((currentConfig, key) => {
      currentConfig[key] = options.screenshotConfig[key];
      return currentConfig;
    }, merge({}, DEFAULT_SCREENSHOT_CONFIG, getConfig().screenshotConfig));

  screenshotConfig.blackout = screenshotConfig.blackout || [];
  screenshotConfig.blackout.push('.snapshot-diff');
  return screenshotConfig;
}

function getCustomName(suppliedConfig) {
  const cfg = suppliedConfig || getConfig();
  return cfg.name;
}

function getCustomSeparator(suppliedConfig) {
  const cfg = suppliedConfig || getConfig();
  return cfg.separator;
}

function getServerUrl(suppliedConfig) {
  const cfg = suppliedConfig || getConfig();
  return `http://${cfg.serverHost}:${cfg.serverPort}/?token=${cfg.token}`;
}

function shouldNormalize(dataType, suppliedConfig) {
  const cfg =
    suppliedConfig && suppliedConfig.normalizeJson !== undefined ? suppliedConfig : getConfig();
  return dataType === TYPE_JSON && cfg.normalizeJson;
}

function getPrettierConfig(dataType, suppliedConfig) {
  const cfg = suppliedConfig && suppliedConfig.prettierConfig ? suppliedConfig : getConfig();
  return cfg.prettier && cfg.prettierConfig ? cfg.prettierConfig[dataType] : undefined;
}

module.exports = {
  CONFIG_KEY,
  DEFAULT_IMAGE_CONFIG,
  DEFAULT_SCREENSHOT_CONFIG,
  createToken,
  getConfig,
  getImageConfig,
  getPrettierConfig,
  getScreenshotConfig,
  getCustomName,
  getCustomSeparator,
  getServerUrl,
  initConfig,
  shouldNormalize,
};
